src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/RazorLSPHandlerBase.cs

```diff
@@ -0,0 +1,151 @@
+            string method,
+            LanguageServerKind serverKind,
+            TIn parameters,
+            CancellationToken cancellationToken)
+        {
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (string.IsNullOrEmpty(method))
+            {
+                throw new ArgumentException("message", nameof(method));
+            }
+
+            var contentType = RazorLSPContentTypeDefinition.Name;
+            if (serverKind == LanguageServerKind.CSharp)
+            {
+                contentType = CSharpVirtualDocumentFactory.CSharpLSPContentTypeName;
+            }
+            else if (serverKind == LanguageServerKind.Html)
+            {
+                contentType = HtmlVirtualDocumentFactory.HtmlLSPContentTypeName;
+            }
+
+            return languageClientBroker.RequestAsync<TIn, TOut>(
+                new[] { contentType },
+                cap => true,
+                method,
+                parameters,
+                cancellationToken);
+        }
+
+        protected virtual async Task<ProjectionResult> GetProjectionAsync(LSPDocumentSnapshot documentSnapshot, Position position, CancellationToken cancellationToken)
+        {
+            var languageQueryParams = new RazorLanguageQueryParams()
+            {
+                Position = new OmniSharpPosition(position.Line, position.Character),
+                Uri = documentSnapshot.Uri
+            };
+
+            var languageResponse = await RequestServerAsync<RazorLanguageQueryParams, RazorLanguageQueryResponse>(
+                LanguageClientBroker,
+                LanguageServerConstants.RazorLanguageQueryEndpoint,
+                LanguageServerKind.Razor,
+                languageQueryParams,
+                cancellationToken);
+
+            VirtualDocumentSnapshot virtualDocument;
+            if (languageResponse.Kind == RazorLanguageKind.CSharp &&
+                documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDoc))
+            {
+                virtualDocument = csharpDoc;
+            }
+            else if (languageResponse.Kind == RazorLanguageKind.Html &&
+                documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlDoc))
+            {
+                virtualDocument = htmlDoc;
+            }
+            else
+            {
+                return null;
+            }
+
+            if (languageResponse.HostDocumentVersion == UndefinedDocumentVersion)
+            {
+                // There should always be a document version attached to an open document.
+                // TODO: Log it and move on as if it was synchronized.
+            }
+            else
+            {
+                var synchronized = await DocumentSynchronizer.TrySynchronizeVirtualDocumentAsync(documentSnapshot, virtualDocument, cancellationToken);
```

Comment:
Prior to doing this you'll want to verify that the `langaugeResponse.HostDocumentVersion` matches `documentSnapshot.Version`. I assume it always will but if it doesn't we may need to expand the synchronizer to wait for new document snapshots.

If it doesn't you may want to Debug.Fail so we can check if our assumptions are correct.